storage: TestStoreRangeRemoveDead failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/eb8345b19cf6c15b3e4fcb9c156136357d83cb2d

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1016077&tab=buildLog

```
I181116 10:15:34.390465 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:35.393431 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:36.395719 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:37.416681 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:38.425534 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:39.428573 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:40.430537 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:41.433634 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:42.447051 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:43.449393 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:44.451628 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:45.458339 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:46.466393 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:47.475721 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:48.483535 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:49.486071 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:50.499470 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:51.506466 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:52.509391 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:53.511787 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:54.514058 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:55.520339 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:56.524981 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:57.527247 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:58.529486 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
I181116 10:15:59.532671 72590 storage/client_raft_test.go:2907  SucceedsSoon: expected 842353710736 replicas; have [(n1,s1):1]
--- FAIL: TestStoreRangeRemoveDead (46.01s)
	soon.go:49: condition failed to evaluate within 45s: expected 842353710736 replicas; have [(n1,s1):1]
		goroutine 72590 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc426201f20, 0x3e5c060)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3ec2fc0, 0xc420b9c0f0, 0xc421aa6f60)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc420b9c0f0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2907 +0x669
		testing.tRunner(0xc420b9c0f0, 0x36dd0f0)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

Ok, the weird number is due to printing a pointer via `%+v`. The other thing that maybe changed recently is that this test is now running under a default zone config that has 5 replicas (up from 3 a while back). Possibly that is pushing it over the limit. Here's a real failure:

```
--- FAIL: TestStoreRangeRemoveDead (46.16s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n1,s1):1]
		goroutine 22 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc4212fa810, 0x3e5e240)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3ec51a0, 0xc4201d5590, 0xc4203865e0)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4201d5590)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2910 +0x76b
		testing.tRunner(0xc4201d5590, 0x36dee30)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565
```

I'll at the very least send a patch to fix that message. I'm also stressracing this with less parallelism to check if that's at fault for the timeout firing here.

SHA: https://github.com/cockroachdb/cockroach/commits/ce35979e2d8d7fdbffa6543fb646fadbdf442ab3

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1035312&tab=buildLog

```
E181201 08:40:12.267056 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.320520 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.337821 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.355956 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.421598 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.443629 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.466750 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.480485 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.501493 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.523043 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.557726 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.573423 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.619272 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.650595 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.686434 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181201 08:40:12.718623 73566 storage/queue.go:837  [replicate,s4,r1/7:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
I181201 08:40:12.723086 73559 storage/compactor/compactor.go:330  [s5,compactor] purging suggested compaction for range /Min - /Max that contains live data
W181201 08:40:12.731794 73551 storage/store.go:1458  [s5,r1/25:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181201 08:40:12.832501 73724 storage/raft_transport.go:584  while processing outgoing Raft queue to node 6: rpc error: code = Unavailable desc = transport is closing:
W181201 08:40:12.841215 73363 gossip/gossip.go:1501  [n3] no incoming or outgoing connections
W181201 08:40:12.841408 73581 gossip/gossip.go:1501  [n6] no incoming or outgoing connections
--- FAIL: TestStoreRangeRemoveDead (46.38s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n3,s3):26 (n4,s4):7 (n5,s5):25]
		goroutine 73003 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc421307530, 0x3ea6b80)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3f0e4e0, 0xc4201fb950, 0xc42a2c60a0)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4201fb950)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2907 +0x669
		testing.tRunner(0xc4201fb950, 0x37203f0)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

SHA: https://github.com/cockroachdb/cockroach/commits/63f5dc75d0c69950ce0dfb4e7ef3f2e2be2889b9

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1036079&tab=buildLog

```
E181202 07:08:39.312009 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.320330 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.373364 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.385151 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.423736 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.436124 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.451294 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.460104 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.468920 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.482331 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.488260 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.496788 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.505497 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.511893 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.529208 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.535171 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
I181202 07:08:39.536625 73457 storage/compactor/compactor.go:330  [s4,compactor] purging suggested compaction for range /Min - /Max that contains live data
E181202 07:08:39.546181 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.554706 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
E181202 07:08:39.562666 73749 storage/queue.go:837  [replicate,s1,r1/1:/M{in-ax}] 1 matching stores are currently throttled
W181202 07:08:39.582684 73779 storage/raft_transport.go:282  unable to accept Raft message from (n5,s5):?: no handler registered for (n1,s1):?
W181202 07:08:39.584352 73770 storage/store.go:3645  [s5] raft error: node 1 claims to not contain store 1 for replica (n1,s1):?: store 1 was not found
W181202 07:08:39.585487 73768 storage/raft_transport.go:584  while processing outgoing Raft queue to node 1: store 1 was not found:
W181202 07:08:39.598847 73517 storage/store.go:1458  [s5,r1/22:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181202 07:08:39.640361 73703 storage/store.go:1458  [s6,r1/25:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W181202 07:08:39.674292 73361 gossip/gossip.go:1501  [n4] no incoming or outgoing connections
--- FAIL: TestStoreRangeRemoveDead (46.04s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n1,s1):1 (n5,s5):22 (n4,s4):24]
		goroutine 72969 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc421a776b0, 0x3ea6b80)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3f0e4e0, 0xc4207982d0, 0xc421789cc0)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4207982d0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2907 +0x669
		testing.tRunner(0xc4207982d0, 0x37203f0)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

SHA: https://github.com/cockroachdb/cockroach/commits/e6cb0c5c329617b560eee37527248171b5e06382

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1039239&tab=buildLog

```
E181204 07:45:36.423437 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.446065 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.454167 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.487695 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.521933 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.564676 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.582888 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.590020 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.634574 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
E181204 07:45:36.687593 73675 storage/queue.go:837  [replicate,s6,r1/11:/M{in-ax}] avoid up-replicating to fragile quorum: 1 matching stores are currently throttled
W181204 07:45:36.701576 73797 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (server shutdown)
I181204 07:45:36.701812 73797 storage/node_liveness.go:790  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (server shutdown)
W181204 07:45:36.704120 73797 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W181204 07:45:36.704636 73797 storage/node_liveness.go:454  [hb] failed node liveness heartbeat: node unavailable; try another peer
W181204 07:45:36.719018 74011 storage/raft_transport.go:282  unable to accept Raft message from (n6,s6):11: no handler registered for (n1,s1):27
I181204 07:45:36.723411 73823 storage/compactor/compactor.go:330  [s5,compactor] purging suggested compaction for range /Min - /Max that contains live data
W181204 07:45:36.753128 74011 storage/raft_transport.go:282  unable to accept Raft message from (n6,s6):11: no handler registered for (n1,s1):27
W181204 07:45:36.754022 74011 storage/raft_transport.go:282  unable to accept Raft message from (n6,s6):?: no handler registered for (n1,s1):?
W181204 07:45:36.760759 73793 storage/store.go:3645  [s6,r1/11:/M{in-ax}] raft error: node 1 claims to not contain store 1 for replica (n1,s1):27: store 1 was not found
W181204 07:45:36.772222 74054 storage/raft_transport.go:584  while processing outgoing Raft queue to node 1: store 1 was not found:
W181204 07:45:36.829405 74040 storage/raft_transport.go:584  while processing outgoing Raft queue to node 3: EOF:
--- FAIL: TestStoreRangeRemoveDead (46.48s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n4,s4):26 (n6,s6):11 (n5,s5):25]
		goroutine 73217 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc4213d5bf0, 0x3ea7420)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3f0eda0, 0xc4211c6000, 0xc4232e7a00)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4211c6000)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2907 +0x669
		testing.tRunner(0xc4211c6000, 0x3720b98)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -2860,20 +2860,30 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 	sc := storage.TestStoreConfig(nil)
 	sc.TestingKnobs.DisableReplicaRebalancing = true
 	mtc := &multiTestContext{storeConfig: &sc}
-	mtc.timeUntilStoreDead = storage.TestTimeUntilStoreDead
-	defer mtc.Stop()
+
+	storage.TimeUntilStoreDead.Override(&sc.Settings.SV, storage.TestTimeUntilStoreDead)
+	// Disable declined and failed reservations. Their default values are on
+	// the order of seconds whereas we'll only advance the manual clock used
+	// in this test very slowly, leading to timeouts otherwise.
+	storage.DeclinedReservationsTimeout.Override(&sc.Settings.SV, time.Nanosecond)
+	storage.FailedReservationsTimeout.Override(&sc.Settings.SV, time.Nanosecond)
 
 	zone := config.DefaultSystemZoneConfig()
 	mtc.Start(t, int(*zone.NumReplicas+1))
+	defer mtc.Stop()
 
 	var nonDeadStores []*storage.Store
+	var deadStore *storage.Store
 	for i, s := range mtc.stores {
 		if i == 1 {
-			// Skip the dead store.
+			deadStore = s
+
 			continue
 		}
 		nonDeadStores = append(nonDeadStores, s)
 	}
+
+	var dead int32 // atomically
 
 	// Create a goroutine to gossip store capacity info periodically.
 	go func() {
@@ -2885,9 +2895,18 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 			select {
 			case <-ticker.C:
 				mtc.manualClock.Increment(int64(tickerDur))
-
+				storeIsDead := atomic.LoadInt32(&dead) == 1
 				// Keep gossiping the stores, excepting the dead store.
-				for _, s := range nonDeadStores {
+				n := len(nonDeadStores)
+				if !storeIsDead {
+					// Until the deadStore dies, don't gossip one of the other
+					// stores to force some replicas on deadStore.
+					n--
+					if err := deadStore.GossipStore(context.Background(), false /* useCached */); err != nil {
+						panic(err)
+					}
+				}
+				for _, s := range nonDeadStores[:n] {
 					if err := s.GossipStore(context.Background(), false /* useCached */); err != nil {
 						panic(err)
 					}
@@ -2903,19 +2922,28 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 		}
 	}()
 
-	// Wait for up-replication.
+	// Wait for up-replication, including at least one replica on s2 (deadStore).
 	testutils.SucceedsSoon(t, func() error {
 		replicas := getRangeMetadata(roachpb.RKeyMin, mtc, t).Replicas
-		if len(replicas) == int(*zone.NumReplicas) {
-			return nil
+		var found bool
+		for _, replica := range replicas {
+			if replica.StoreID == deadStore.StoreID() {
+				found = true
+			}
 		}
-		return errors.Errorf("expected %d replicas; have %+v", *zone.NumReplicas, replicas)
+		if len(replicas) != int(*zone.NumReplicas) {
+			return errors.Errorf("expected %d replicas; have %+v", *zone.NumReplicas, replicas)
+		}
+		if !found {
+			return errors.Errorf("no replica found on s%d", deadStore.StoreID())
+		}
+		return nil
 	})
 
 	// Stop a store which will be rebalanced away from. We can't use the very first
 	// one since getRangeMetadata is hard-coded to query that one, so we'll use the
 	// one after that.
-	deadStoreID := mtc.stores[1].StoreID() // 2
+	atomic.StoreInt32(&dead, 1)
 	mtc.stopStore(1)
 	// The mtc asks us to restart this store before stopping the mtc.
 	defer mtc.restartStoreWithoutHeartbeat(1)
@@ -2926,7 +2954,7 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 			return errors.Errorf("expected %d replicas; have %+v", zone.NumReplicas, replicas)
 		}
 		for _, r := range replicas {
-			if r.StoreID == deadStoreID {
+			if r.StoreID == deadStore.StoreID() /* 2 */ {
 				return errors.Errorf("expected store %d to be replaced; have %+v", r.StoreID, replicas)
 			}
 		}

--- a/pkg/storage/client_test.go
+++ b/pkg/storage/client_test.go
@@ -207,9 +207,8 @@ type multiTestContext struct {
 	// We use multiple stoppers so we can restart different parts of the
 	// test individually. transportStopper is for 'transport', and the
 	// 'stoppers' slice corresponds to the 'stores'.
-	transportStopper   *stop.Stopper
-	engineStoppers     []*stop.Stopper
-	timeUntilStoreDead time.Duration
+	transportStopper *stop.Stopper
+	engineStoppers   []*stop.Stopper
 
 	// The fields below may mutate at runtime so the pointers they contain are
 	// protected by 'mu'.
@@ -243,7 +242,6 @@ func (m *multiTestContext) Start(t testing.TB, numStores int) {
 		mCopy.engines = nil
 		mCopy.engineStoppers = nil
 		mCopy.injEngines = false
-		mCopy.timeUntilStoreDead = 0
 		var empty multiTestContext
 		if !reflect.DeepEqual(empty, mCopy) {
 			t.Fatalf("illegal fields set in multiTestContext:\n%s", pretty.Diff(empty, mCopy))
@@ -678,11 +676,12 @@ func (m *multiTestContext) populateDB(idx int, stopper *stop.Stopper) {
 	m.dbs[idx] = client.NewDB(ambient, tcsFactory, m.clocks[idx])
 }
 
-func (m *multiTestContext) populateStorePool(idx int, nodeLiveness *storage.NodeLiveness) {
-	storage.TimeUntilStoreDead.Override(&m.storeConfig.Settings.SV, m.timeUntilStoreDead)
+func (m *multiTestContext) populateStorePool(
+	idx int, cfg storage.StoreConfig, nodeLiveness *storage.NodeLiveness,
+) {
 	m.storePools[idx] = storage.NewStorePool(
-		m.storeConfig.AmbientCtx,
-		m.storeConfig.Settings,
+		cfg.AmbientCtx,
+		cfg.Settings,
 		m.gossips[idx],
 		m.clocks[idx],
 		storage.MakeStorePoolNodeLivenessFunc(nodeLiveness),
@@ -741,9 +740,6 @@ func (m *multiTestContext) addStore(idx int) {
 		m.transportStopper,
 		metric.NewRegistry(),
 	)
-	if m.timeUntilStoreDead == 0 {
-		m.timeUntilStoreDead = storage.TestTimeUntilStoreDeadOff
-	}
 
 	nodeID := roachpb.NodeID(idx + 1)
 	cfg := m.makeStoreConfig(idx)
@@ -754,7 +750,7 @@ func (m *multiTestContext) addStore(idx int) {
 		ambient, m.clocks[idx], m.dbs[idx], m.engines, m.gossips[idx],
 		nlActive, nlRenewal, cfg.Settings, metric.TestSampleInterval,
 	)
-	m.populateStorePool(idx, m.nodeLivenesses[idx])
+	m.populateStorePool(idx, cfg, m.nodeLivenesses[idx])
 	cfg.DB = m.dbs[idx]
 	cfg.NodeLiveness = m.nodeLivenesses[idx]
 	cfg.StorePool = m.storePools[idx]
@@ -913,7 +909,7 @@ func (m *multiTestContext) restartStoreWithoutHeartbeat(i int) {
 		log.AmbientContext{Tracer: m.storeConfig.Settings.Tracer}, m.clocks[i], m.dbs[i], m.engines,
 		m.gossips[i], nlActive, nlRenewal, cfg.Settings, metric.TestSampleInterval,
 	)
-	m.populateStorePool(i, m.nodeLivenesses[i])
+	m.populateStorePool(i, cfg, m.nodeLivenesses[i])
 	cfg.DB = m.dbs[i]
 	cfg.NodeLiveness = m.nodeLivenesses[i]
 	cfg.StorePool = m.storePools[i]

--- a/pkg/storage/store_pool.go
+++ b/pkg/storage/store_pool.go
@@ -45,15 +45,19 @@ const (
 	TestTimeUntilStoreDeadOff = 24 * time.Hour
 )
 
-// declinedReservationsTimeout needs to be non-zero to prevent useless retries
-// in the replicateQueue.process() retry loop.
-var declinedReservationsTimeout = settings.RegisterNonNegativeDurationSetting(
+// DeclinedReservationsTimeout specifies a duration during which the local
+// replicate queue will not consider stores which have rejected a reservation a
+// viable target.
+var DeclinedReservationsTimeout = settings.RegisterNonNegativeDurationSetting(
 	"server.declined_reservation_timeout",
 	"the amount of time to consider the store throttled for up-replication after a reservation was declined",
 	1*time.Second,
 )
 
-var failedReservationsTimeout = settings.RegisterNonNegativeDurationSetting(
+// FailedReservationsTimeout specifies a duration during which the local
+// replicate queue will not consider stores which have failed a reservation a
+// viable target.
+var FailedReservationsTimeout = settings.RegisterNonNegativeDurationSetting(
 	"server.failed_reservation_timeout",
 	"the amount of time to consider the store throttled for up-replication after a failed reservation call",
 	5*time.Second,
@@ -715,7 +719,7 @@ func (sp *StorePool) throttle(reason throttleReason, storeID roachpb.StoreID) {
 	// timeout period has passed.
 	switch reason {
 	case throttleDeclined:
-		timeout := declinedReservationsTimeout.Get(&sp.st.SV)
+		timeout := DeclinedReservationsTimeout.Get(&sp.st.SV)
 		detail.throttledUntil = sp.clock.PhysicalTime().Add(timeout)
 		if log.V(2) {
 			ctx := sp.AnnotateCtx(context.TODO())
@@ -723,7 +727,7 @@ func (sp *StorePool) throttle(reason throttleReason, storeID roachpb.StoreID) {
 				storeID, timeout, detail.throttledUntil)
 		}
 	case throttleFailed:
-		timeout := failedReservationsTimeout.Get(&sp.st.SV)
+		timeout := FailedReservationsTimeout.Get(&sp.st.SV)
 		detail.throttledUntil = sp.clock.PhysicalTime().Add(timeout)
 		if log.V(2) {
 			ctx := sp.AnnotateCtx(context.TODO())

--- a/pkg/storage/store_pool_test.go
+++ b/pkg/storage/store_pool_test.go
@@ -755,7 +755,7 @@ func TestStorePoolThrottle(t *testing.T) {
 	sg.GossipStores(uniqueStore, t)
 
 	{
-		expected := sp.clock.Now().GoTime().Add(declinedReservationsTimeout.Get(&sp.st.SV))
+		expected := sp.clock.Now().GoTime().Add(DeclinedReservationsTimeout.Get(&sp.st.SV))
 		sp.throttle(throttleDeclined, 1)
 
 		sp.detailsMu.Lock()
@@ -768,7 +768,7 @@ func TestStorePoolThrottle(t *testing.T) {
 	}
 
 	{
-		expected := sp.clock.Now().GoTime().Add(failedReservationsTimeout.Get(&sp.st.SV))
+		expected := sp.clock.Now().GoTime().Add(FailedReservationsTimeout.Get(&sp.st.SV))
 		sp.throttle(throttleFailed, 1)
 
 		sp.detailsMu.Lock()

